### ISSUE TITLE
@const attribute

### ISSUE THREAD

**Issue by k2d222** [wgsl-tooling-wg/wesl-spec#46]
WGSL defines the [@const](https://www.w3.org/TR/WGSL/#const-attr) attribute on functions: 

https://www.w3.org/TR/WGSL/#const-functions
> A function declared with a [const](https://www.w3.org/TR/WGSL/#attribute-const) attribute can be evaluated at [shader-creation time](https://www.w3.org/TR/WGSL/#shader-module-creation). These functions are called const-functions. Calls to these functions can be part of [const-expressions](https://www.w3.org/TR/WGSL/#const-expressions).

Sadly (but understandably), these attributes cannot be written in user code. I wonder if we should spec it as an *optional* WESL extension. They are very difficult to implement (requires implementing both `eval()` for Expressions and `exec()` for declarations), but my implementation supports it already.

**Comment by stefnotch** [comment]
Is there an implementation option where one just naively inlines such functions? Or does that only cover expressions, and not general functions with loops and whatnot?

**Comment by k2d222** [comment]
> Is there an implementation option where one just naively inlines such functions? Or does that only cover expressions, and not general functions with loops and whatnot?

Right, that could be a possibility, I don't see any reason for it to behave differently.

**Comment by mighdoll** [comment]
are you suggesting making this optional because eval() and exec() are difficult to implement in a small runtime linker? 

I get the utility of user supplied @const functions in principle, but are there specific example use cases you have in mind that to enable?

**Comment by k2d222** [comment]
@stefnotch 
so I just realized that inlining would not work if the function call is a `CallExpression` (not a `CallStatement`). :disappointed: 

@mighdoll 
Yes, IMO they would be optional if they are difficult to implement. That being said, there are a few, very good reasons to have them.

### Example 1: say we adopt Use.GPU's technique of using overridable getters (the `@link` syntax):
```rs
@const getArrayLength() { return 1u; } // is overridable and provides a default
@const getFooBinding() { return 1u; }  // is overridable and provides a default
// ...
@group(0) @binding(getFooBinding()) foo: u32;
let myArray: array<u32, getArrayLength()>;
```

### Example 2: compile-time assertions (e.g. to catch invalid shader permutations)
```rs
@const fn is_foo_valid(myFoo: Foo) -> bool { // this function can also be used at runtime
   return myFoo.x < myFoo.y && sqrt(myFoo.z) < 3.0;
}
// ...
const_assert is_foo_valid();

// ... alternatively, write the const_asserts directly in the function body
@const fn assert_foo_valid(myFoo: Foo) { // but this function cannot be used at runtime
   const_assert myFoo.x < myFoo.y;
   const_assert sqrt(myFoo.z) < 3.0;
}
```

### Example 3: catching [indeterminate values](https://www.w3.org/TR/WGSL/#indeterminate-values)

Some WGSL constructs can produce Undefined Behavior (called indeterminate values in the spec). At runtime, it is just not determined what the value should be. But at compile-time, it is a shader-creation error. 
```rs
let v: vec3f;
let f: f32 = 5.0;
let i: i32 = 5;
v[4] // this is allowed at runtime (same for matrices and arrays) and returns an indeterminate f32 in this case.

// the following operations lead to IEEE-754 errors (yield NaN or infinity). they are UB at runtime but caught at shader-creation.
f / 0.0; // infinity
f = 3.40282347E38 + 1.0 // float overflow
i = 2147483647 + 1 // integer overflow
sqrt(-1)  // invalid operation. Applies to these functions too: acos, acosh, asin, atanh, cos, distance, inverseSqrt, log, log2, normalize, pow, sin, sqrt, tan
```

**Comment by mighdoll** [comment]
Excellent to have examples!

I'm not sure about making array sizes polymorphic in example 1. Array sizes can't really be anything but constants in wgsl, they're compiled into the shader, right? They can't be written as uniform accesses, for example. So I'm not sure it'll make to have a polymorphic data abstraction over that case.

Example 2 - shows how allow validation functions for const_assert could be shared with a runtime test. I think it'd also be nice for factoring validation functions across multiple asserts. The way the WGSL spec is written, it seems like they had this in mind I wonder if there's an issue.

Example 3 I don't understand, where's the @const? Maybe you're suggesting these indeterminate values are more situtations where you'd want a const_assert validation function like in Example 2?

**Comment by stefnotch** [comment]
> @stefnotch so I just realized that inlining would not work if the function call is a `CallExpression` (not a `CallStatement`). 😞
> 

I don't quite understand what you mean with this. Example please?

**Comment by k2d222** [comment]
re @mighdoll 

re example 1: The array size is a constant, because its expression is a const-expression, because it calls a `@const` function.
In wgsl you can already provide an arbitrary const-expression in place of array size.
But tbh this example is a bit contrieved because it might make more sense to `@link` a constant instead of a function. e.g. `@link @const arrayLength = 1u; // can be overriden`. I don't think Use.GPU can do that currently?

re example 2: yeah this shows basically that I would like to perform as much static checking as possible. makes it easier to catch misuses.

re example 3: I forgot to mention an important information here. The point is to run these faillible operations in statements. There is no way to have const-evaluated statements in vanilla wgsl.

Oh and it just made me think of another, even better use-case. If you can use const statements you can use const control flow.
```rs
@const fn primaryColor(theme: u32) {
  switch theme {
    case 0u: return blue;
    case 1u: return red;
    default: return green;
  }
}

const COLOR_THEME: u32 = 0u; // a constant enum (0 = dark theme, 1 = light, 2 = contrast...)
const PRIMARY: vec4 = primaryColor(COLOR_THEME);
```

here's one that catches IEEE-754 errors, with a div and a sqrt. I'm just making random math here. What are the valid values of p? no idea, it's difficult to reason about.
```rs
@const computeSomeConst(p: f32) -> f32 {
  var p2 = p;
  var k = 5.0;
  for (var i = 0; i < i32(sqrt(p2 + k)); i++) {
    k /= p;
    p -= 0.5;
  }
  return k;
}

const MY_CONST = ComputeSomeConst(SOME_PARAM);
```

I'll try to provide a better example later that uses all of these use-cases.

**Comment by k2d222** [comment]
> > @stefnotch so I just realized that inlining would not work if the function call is a `CallExpression` (not a `CallStatement`). 😞
> > 
> 
> I don't quite understand what you mean with this. Example please?

Sure! in WGSL there's actually two ways to call functions: in expressions and in statements. And naturally, void functions can only be called in statements.

```rs
@const fn my_func(ptr<u32>) -> u32 { return 5u; }

let x = my_func();
// cannot be inlined, it would be let x = { return 5u; }

let p: u32;
my_func(*p);
// can be inlined
```
hope this clarifies.

**Comment by stefnotch** [comment]
I would have expected the first one to turn into `let x = 5u`. That's a more involved rewriting action though.

And yes, like you have noticed, we're pretty much limited to functions that are a single expression if we go down the road of inlining. No `if`, no `switch` nor any loops.
